Sources/SwiftSyntax/generated/SyntaxRewriter.swift

```diff
@@ -3898,79 +3909,52 @@ open class SyntaxRewriter {
 
-      let rewritten = dispatchVisit(Syntax(absoluteRaw, parent: syntaxNode))
-      if rewritten.id != info.nodeId {
+      dispatchVisit(&childNode)
+      if childNode.raw.id != child.id {
         // The node was rewritten, let's handle it
-        if newLayout == nil {
-          // We have not yet collected any previous rewritten nodes. Initialize
-          // the new layout with the previous nodes of the parent. This is
-          // possible, since we know they were not rewritten.
 
-          // The below implementation is based on Collection.map but directly
-          // reserves enough capacity for the entire layout.
-          newLayout = ContiguousArray<RawSyntax?>()
-          newLayout!.reserveCapacity(node.raw.layoutView!.children.count)
-          for j in 0 ..< childIndex {
-            newLayout!.append(node.raw.layoutView!.children[j])
-          }
+        if newLayout.baseAddress == nil {
+          // We have not yet collected any previous rewritten nodes. Initialize
+          // the new layout with the previous nodes of the parent.
+          newLayout = .allocate(capacity: node.raw.layoutView!.children.count)
+          _ = newLayout.initialize(fromContentsOf: node.raw.layoutView!.children)
         }
 
-        // Now that we know we have a new layout in which we collect rewritten
-        // nodes, add it.
-        rewrittens.append(rewritten)
-        newLayout!.append(rewritten.raw)
-      } else {
-        // The node was not changed by the rewriter. Only store it if a previous
-        // node has been rewritten and we are collecting a rewritten layout.
-        if newLayout != nil {
-          newLayout!.append(raw)
-        }
+        // Update the rewritten child.
+        newLayout[Int(info.indexInParent)] = childNode.raw
+        // Retain the syntax arena of the new node until it's wrapped with Syntax node.
+        rewrittens.append(childNode.raw.arenaReference.retained)
       }
+
+      // Recycle 'childNode.info'
+      nodeFactory.dispose(&childNode)
     }
 
-    if let newLayout {
+    if newLayout.baseAddress != nil {
       // A child node was rewritten. Build the updated node.
 
-      // Sanity check, ensure the new children are the same length.
-      precondition(newLayout.count == node.raw.layoutView!.children.count)
-
       let arena = self.arena ?? SyntaxArena()
-      let newRaw = node.raw.layoutView!.replacingLayout(with: Array(newLayout), arena: arena)
+      let newRaw = node.raw.layoutView!.replacingLayout(with: newLayout, arena: arena)
+      newLayout.deinitialize()
+      newLayout.deallocate()
       // 'withExtendedLifetime' to keep 'SyntaxArena's of them alive until here.
       return withExtendedLifetime(rewrittens) {
-        Syntax(raw: newRaw, rawNodeArena: arena).cast(SyntaxType.self)
+        Syntax(raw: newRaw, rawNodeArena: arena)
       }
     } else {
       // No child node was rewritten. So no need to change this node as well.
-      return node
+      return nil
```

Comment:
Is there a benefit from returning `nil` vs `node` here? AFAICT we always to `?? node` on every call and we could avoid that if `visitChildren` didn’t return an optional.

Reply:
Oh true, will try.

Reply:
Ah e.g. `visitChildren(node._syntaxNode)?.cast(AccessorEffectSpecifiersSyntax.self) ?? node`
For this pattern, `.cast(AccessorEffectSpecifiersSyntax.self)` have some cost (even after `__shared` change). So being able to use `node` as-is was faster

Reply:
Sorry I re-measured and I don't see any difference returning `node` instead of `nil`. So I changed it to return `node` and remove `?? node` in the caller.